### PR TITLE
injector changes for health check, pilot agent take over app readiness check.

### PR DESCRIPTION
**Pull request by incfly**
Injector part for https://github.com/istio/istio/issues/9150

- Figure out the application probing path and port and fill that in the sidecar tmpl.
- Rewrite the app liveness and readiness probe.

The part to figure out the `statusPort` is ugly, but considering the sidecartmpl can come from configmap, I don't think there's a better way. Feel free to comment.

Won't merge util https://github.com/istio/istio/pull/9204 is merged. Will add a E2E test before merging as well.

### CODE REVIEW

**Comment by incfly** [comment]
/hold

Add hold just to address e2e test.

**Comment by wattli** [comment]
Any update?

**Comment by incfly** [comment]
@wattli I got the basic example working, readiness health check work under mTLS enabled for a pod. But still need some polishing for some corner cases handling.

**Comment by incfly** [comment]
/hold cancel

While the tests needs to be fixed, I think this PR is ready for review. PTAL @mandarjog @ayj thanks!

**Comment by hzxuzhonghu** [comment]
@incfly I have a doubt: when have multi-containers, and each has a readiness probe. 
Seems `appReadyUrl` only support one url redirect.

**Comment by hzxuzhonghu** [comment]
Another thought:  

1. remove `--appReadyUrl` and `--appLiveUrl` and only rewrite the probe port with 10250.

1. in pilot-agent, status server uses a wilcard prefix "/" match. And acquire `req.RequestURI` in handler. Use it as the probe path

1. The only issue left is how we can get the probe port, we can try each port in `--applicationPorts`.

I think this way is easy for users.

File: pilot/pkg/kube/inject/app_probe.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	StatusPortCmdFlagName = "statusPort"
+
+	istioProxyContainerName = "istio-proxy"
+)
+
+func appProbePath(kind string, containers []corev1.Container) string {
+	if !canRewriteProber(containers) {
+		return ""
+	}
+	for _, c := range containers {
```
Comment (hzxuzhonghu) [inline pilot/pkg/kube/inject/app_probe.go]: If more than one containers with readiness or liveness probe, then this can not work well.

**Comment by incfly** [comment]
> Another thought:
> 
> 1. remove `--appReadyUrl` and `--appLiveUrl` and only rewrite the probe port with 10250.
> 2. in pilot-agent, status server uses a wilcard prefix "/" match. And acquire `req.RequestURI` in handler. Use it as the probe path
> 3. The only issue left is how we can get the probe port, we can try each port in `--applicationPorts`.
> 
> I think this way is easy for users.

Great point. I like the idea of rewriting probe port with 10502 directly, instead of doing encoding of multi containers url paths into `--appReadyUrl`.

However, I feel we don't want to enumerate all ports for the same path from applicationPorts, because that subjects to ambiguity, for example, if port 8000 fails and 8001 succeeds, what should pilot agent think? We have two options to avoid that
  - Put a http header in [`httpHeader` sections](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-liveness-http-request) indicate this, using reserved header name, saying "appProbePort: 8000"
  - Rewrite url path to encode appProbe port as well. Could be `:15020/8000/original_probe_path`

Comments? @mandarjog @hzxuzhonghu

**Comment by hzxuzhonghu** [comment]
>Rewrite url path to encode appProbe port as well. Could be :15020/8000/original_probe_path

I like this, more clear.

File: pilot/pkg/kube/inject/app_probe.go (hunk 1 above)
Comment (incfly) [reply pilot/pkg/kube/inject/app_probe.go]: This should work now by changing to header indicating the app port.

File: pilot/pkg/kube/inject/app_probe.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+func rewriteAppHTTPProbe(spec *SidecarInjectionSpec, podSpec *corev1.PodSpec) {
+	statusPort := -1
+	pi := -1
+	for _, c := range spec.Containers {
+		if c.Name != istioProxyContainerName {
+			continue
+		}
+		for i, arg := range c.Args {
+			// arg is "--flag-name"
```
[the remark this answers is not in the document]
Comment (incfly) [reply pilot/pkg/kube/inject/app_probe.go]: Thanks for the suggestions. Applied the changes. Only difference is that I instead return the corner case first to reduce the nest level.

File: pilot/pkg/kube/inject/app_probe.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	// We reuse it for taking over application's readiness probing as well.
+	// TODO: replace the hardcoded statusPort elsewhere by this variable as much as possible.
+	StatusPortCmdFlagName = "statusPort"
+
+	// TODO: any constant refers to this container's name?
+	istioProxyContainerName = "istio-proxy"
+)
+
+var (
+	statusPortPattern = regexp.MustCompile(fmt.Sprintf("^-{1,2}%s$", StatusPortCmdFlagName))
```
Comment (hzxuzhonghu) [inline pilot/pkg/kube/inject/app_probe.go]: Should we support matching  `--statusPort=15020` ?

**Comment by hzxuzhonghu** [comment]
/approve

give others a chance to look.

File: pilot/pkg/kube/inject/app_probe.go (hunk 3 above)
Comment (nmittler) [reply pilot/pkg/kube/inject/app_probe.go]: Yes, good catch.

**Comment by nmittler** [comment]
@incfly new logic looks good.  To @hzxuzhonghu's comment, just need to make sure that this works with all flag formats.  I think that would include: "--key value", "--key=value", "-key value", "-key=value".

**Comment by incfly** [comment]
@nmittler I addressed the comments, PTAL, thanks!

**Comment by nmittler** [comment]
@incfly can you add unit tests to verify your regex?

**Comment by incfly** [comment]
@nmittler Sure. I didn't add that originally because `rewriteAppProbe` is a private function, which I feel unit test should not be targeted against in general.

I know we can make `app_probe_test.go` as the same package such it can access the method.

Another way. I also did some search and find in some golang std lib code, people export internal function with a thin wrapper and then test it. e.g https://golang.org/src/net/http/export_test.go,  https://www.reddit.com/r/golang/comments/9ah8mg/should_we_unit_test_unexported_methods/

Any advice on this?

**Comment by nmittler** [comment]
@incfly I don't feel too strongly here. If you want to export the method or just put the unit test in the same package ... either way is reasonable.

**Comment by incfly** [comment]
Unit test added. @nmittler

**Comment by nmittler** [comment]
@incfly thanks!

/lgtm
/approve

**Comment by hzxuzhonghu** [comment]
@incfly Should this go to release-1.1 since the original part fir pilot-agent  already in.

**Comment by incfly** [comment]
@hzxuzhonghu I'll check the 1.1 branch and if it works, I'll CP this PR to 1.1 branch.

**Pull request merged**